### DISCUSSION TITLE
Higher order accuracies with muliple Ghost values

### DISCUSSION THREAD

**Discussion by shadowbourne** [wenceorg/comp4187#103] [Q&A]
In 2b we are asked to augment the problem to enforce 4th order accurate neumann conditions at the boundary.
We were taught in the lecture course how to enforce lower order accuracy conditions by deriving a stencil and using this to substitute the value of the new ghost value u(n+1) into the main stencil.

However, this methodology we have been taught does not work when there are multiple ghost values in the stencil (u(n+1) and u(n+2)), as in the assignments question (from my understanding). 

Because the derived relation involves two ghost values and a simulataneous equation would have to be set up / solved, however there are too many degrees of freedom to determine the individual values of u(n+1) and u(n+2).

I could find nothing online about such a case, and was wondering what one is supposed to do?

Thanks!

**Comment by annereinarz** [comment]
Your suggested approach seems correct to me. We introduce one (or more) ghost values outside of the domain and use these in order to compute the values of the boundary nodes. The higher the accuracy we want the more points will be in the stencil, meaning that more ghost values are needed. In this case with a fourth order accurate stencil we need to set 2 ghost values. 
With the second order accurate scheme from the lecture we needed only one point, so we set:
𝑢_{𝑛+1}=𝑢_{𝑛−1}+𝑏 2h
If we have a second (or more) point beyond that we can extrapolate further using the same approach, to give a condition of the same form. The main difference is that the node u_{n+2} is not a distance of 2h from the node u_{n-1}.

**Reply by shadowbourne** [reply]
Okay, I think I must have completely misunderstood the theory then.

Are you saying u_{n+2} simply = u{n-1} + b3h ?

 with b = the general H in the question.

If so is there any theory behind why its just u{n-1} + b3h ?

As I thought in lectures u_{n+1} simply = u{n-1} + b2h was obtained by rearranging the centred differences formula u'(n) = (u_{n+1} - u_{n-1})/2h

Therefore:

My methodology was to obtain fourth order accurate approximation for u'(x) (instead of the original 2nd order centred difference stencil):
![image](https://user-images.githubusercontent.com/18665030/167727043-5f28e51c-94e6-42ac-9901-a3808f891f63.png)

To then use to enforce the condition u'(n) = H,
equivalently:
![image](https://user-images.githubusercontent.com/18665030/167727109-0c79e67c-a6cf-4385-a381-a4c28bc21490.png)

Is there any foundation of such an approach or have I simply gotten very confused? 


Thanks!

**Reply by annereinarz** [reply]
Are you saying u_{n+2} simply = u{n-1} + b3h ?
No, that's not quite right. But its quite close. 

Sorry about introducing confusion with b, I was copying from the lecture notes.

You've got one approach right in what you've written there. To follow on from your current approach most easily try to just approximate one step into the domain. Then you'll end up with only one ghost value. I can explain alternative approaches as well, but that's the easiest way to continue on from where you are.

**Reply by shadowbourne** [reply]
u_{n+2} simply = u{n-2} + b4h ? 

When you say approximately one step into the domain do you mean out of the domain or into? ( As in would I ignore u_{n+2} and set u_{n+1} to fit my relation? Or am I misunderstanding)

Would using only one ghost value / approximations step reduce the accuracy order?

If so what are some of these alternative approaches that are possible?

**Reply by annereinarz** [reply]
I'm getting quite close to just writing down the solution, which for obvious reasons I  want to avoid. 

Think about your approach for the Dirichlet conditions. There you presumably modify the stencil for u_0 and u_1 since you need two points outside of the domain. The choice for u_0 is clear, but I'm guessing that you also use a shifted stencil to resolve the u_1 issue? In the sense that you find a 4th order accurate approximation that uses: u_0, u_1, u_2, u_3 and u_4 instead of u_-1, u_0 etc? That's the stencil you want here as well together with your approximation for the first derivative that you use to get a formula for the ghost value u_-1.

Does that make sense? Or did you use a different approach for the Dirichlet values?

Essentially you can construct a 4th order Taylor series expansion out of any nodal values u_k you like. This will give you many different 4th order approximations all of which are valid. It's not something with only one solution. They all have pros and cons.

- If you don't want ghost values at all you would construct a one-sided stencil or u_0 and u_1.
- If you "want" to use more than one ghost value you can extrapolate the solution (also by using a Taylor series of the appropriate order) and then compute using your ghost values. You will end up "mirroring" the interior solution.
